File: cpp/lasagna-master/lasagna_master.cpp
#include "lasagna_master.h"

namespace lasagna_master {

    int preparationTime(std::vector<std::string> layers) {
        return preparationTime(layers, 2);
    }

    int preparationTime(std::vector<std::string> layers, int layerPrepTime=2) {
        return layers.size() * layerPrepTime;
    }

    amount quantities(std::vector<std::string> layers) {
        int noodles = 0;
        double sauce = 0.0;

        for (unsigned int i = 0; i < layers.size(); ++i) {
            if (layers.at(i) == "noodles") {
                noodles += 50;
            }

            if (layers.at(i) == "sauce") {
                sauce += 0.2;
            }
        }

        return amount{ noodles, sauce };
    }

    void addSecretIngredient(std::vector<std::string>& myList, const std::vector<std::string>& friendsList) {
        myList.at(myList.size() - 1) = friendsList.at(friendsList.size() - 1);
    }

    void addSecretIngredient(std::vector<std::string>& myList, std::string secretIngredient) {
        myList.at(myList.size() - 1) = secretIngredient;
    }

    std::vector<double> scaleRecipe(std::vector<double> quantities, int portions) {
        double ratio = static_cast<double>(portions) / 2.0;
        std::vector<double> scaledQuantities{};

        for (unsigned int i = 0; i < quantities.size(); ++i) {
            scaledQuantities.emplace_back(quantities.at(i) * ratio);
        }

        return scaledQuantities;
    }

}  // namespace lasagna_master

File: go/lasagna-master/lasagna_master.go
package lasagna

func PreparationTime(layers []string, timePerLayer int) int {
    if timePerLayer == 0 {
        return 2 * len(layers)
    }

    return timePerLayer * len(layers)
}

func Quantities(layers []string) (int, float64) {
    sauceLayers, noodleLayers := 0, 0
    
    for _, layer := range layers {
        if layer == "sauce" {
            sauceLayers++
        } else if layer == "noodles" {
        	noodleLayers++
        }
    }

    return 50 * noodleLayers, float64(0.2) * float64(sauceLayers)
}

func AddSecretIngredient(friendsList []string, myList []string) {
    myList[len(myList)-1] = friendsList[len(friendsList)-1]
}

func ScaleRecipe(quantities []float64, portions int) []float64 {
    scaledQuantities := make([]float64, 0, len(quantities))

    for _, q := range quantities {
        scaledQuantities = append(scaledQuantities, q * float64(portions) / float64(2))
    }

    return scaledQuantities
}
